README.md
# Timeline

A flutter package that allows you to create basic timelines on your flutter application. This is customizable and easy to plugin to your application. You will find the following features:

* A simple timeline with a title and a description
* Customizable color options for - Line, Title, Description and Background
* Tween animation on the line
* Title will be truncated at 47 characters to prevent overflowing
* Description will be truncated at 75 character to prevent overflowing

<image src="Timeline.png" height="300em"/>

## Getting Started

You will need to add the following dependency in your `pubspec.yaml` file to download the depedency.

```dart
dependencies:
  flutter:
    sdk: flutter
  timeline: 0.0.1
```
Note: This was built on the following tools.

* Flutter - v0.4.4
* Dart - v2.0.0-dev.54.0

### Usage

You will need to create a list of `TimelineModel` first. This creates an instance with an `id` which is a String at the moment, along with `title` and `description`

[Example](https://github.com/rejish4gt/flutter-timeline/blob/master/example/example.dart)

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

## Authors

* **Rejish Radhakrishnan**

## License

Copyright 2018 Rejish Radhakrishnan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## [0.0.1] - Release

* A very basic timeline to display a series of events in the provided order. 

test/timeline_test.dart
import 'package:test/test.dart';

import 'package:timeline/timeline.dart';

void main() {
  test('adds one to input values', () {
    
  });
}


lib/timeline.dart
/* Copyright 2018 Rejish Radhakrishnan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

library timeline;

import 'package:flutter/material.dart';
import 'package:timeline/model/timeline_model.dart';
import 'package:timeline/timeline_element.dart';

class TimelineComponent extends StatefulWidget {

  final List<TimelineModel> timelineList;

  final Color lineColor;

  final Color backgroundColor;

  final Color headingColor;

  final Color descriptionColor;

  const TimelineComponent({Key key, this.timelineList, this.lineColor, this.backgroundColor, this.headingColor, this.descriptionColor}) : super(key: key);

  @override
  TimelineComponentState createState() {
    return new TimelineComponentState();
  }

}

class TimelineComponentState extends State<TimelineComponent> with SingleTickerProviderStateMixin{

  Animation<double> animation;
  AnimationController controller;
  double fraction = 0.0;
 
  @override
  void initState() {
    super.initState();
    controller = AnimationController(
        duration: const Duration(milliseconds: 1000), 
        vsync: this);
    controller.forward();
  }

  @override
  Widget build(BuildContext context) {
    return new Container(
              child: new ListView.builder(
                itemCount: widget.timelineList.length,
                itemBuilder: (_, index) {
                  return new TimelineElement(
                    lineColor: widget.lineColor==null?Theme.of(context).accentColor:widget.lineColor,
                    backgroundColor: widget.backgroundColor==null?Colors.white:widget.backgroundColor,
                    model: widget.timelineList[index],
                    firstElement: index==0,
                    lastElement: widget.timelineList.length==index+1,
                    controller: controller,
                    headingColor: widget.headingColor,
                    descriptionColor: widget.descriptionColor,
                  );
                },
              ),
            );
  }

  @override
    void dispose() {
      controller.dispose();
      super.dispose();
    }

}

example/example.dart
import "package:flutter/material.dart";
import 'package:timeline/model/timeline_model.dart';
import 'package:timeline/timeline.dart';

class TimelineExample extends StatelessWidget {

  final List<TimelineModel> list = [
    TimelineModel(
      id: "1",
      description: "Test 1",
      title: "Test 1"),
      TimelineModel(
      id: "2",
      description: "Test 2",
      title: "Test 2")
  ];

  @override
  Widget build(BuildContext context) {
    return new TimelineComponent(
      timelineList: list,
        // lineColor: Colors.red[200], // Defaults to accent color if not provided
        // backgroundColor: Colors.black87, // Defaults to white if not provided
        // headingColor: Colors.black, // Defaults to black if not provided
        // descriptionColor: Colors.grey, // Defaults to grey if not provided
      );
  }


}

lib/timeline-painter.dart
import 'dart:ui';
import 'package:flutter/material.dart';

class TimelinePainter extends CustomPainter {

  final Color lineColor;
  final Color backgroundColor;
  final bool firstElement;
  final bool lastElement;
  final Animation<double> controller;
  final Animation<double> height;

  TimelinePainter({
    @required this.lineColor,
    @required this.backgroundColor,
    this.firstElement = false,
    this.lastElement = false,
    this.controller
  }) :height = new Tween(begin: 0.0, end: 1.0).animate(new CurvedAnimation(
          parent: controller,
          curve: new Interval(
            0.45, 1.0,
            curve: Curves.ease),
        ),
      ),
      super(repaint: controller);

  @override
  void paint(Canvas canvas, Size size) {
    _centerElementPaint(canvas, size);
  }

  void _centerElementPaint(Canvas canvas, Size size)
  {
    Paint lineStroke = new Paint()
    ..color = lineColor
    ..strokeCap = StrokeCap.round
    ..strokeWidth = 2.0
    ..style = PaintingStyle.stroke;
    if(firstElement)
    {
      Offset offsetCenter = size.center(new Offset(0.0, -4.0));
      Offset offsetBottom = size.bottomCenter(new Offset(0.0, 0.0));
      Offset renderOffset = new Offset(offsetBottom.dx, offsetBottom.dy*(0.5+(controller.value/2)));
      canvas.drawLine(
      offsetCenter, 
      renderOffset, 
      lineStroke);
    }
    else if(lastElement)
    {
      Offset offsetTopCenter = size.topCenter(new Offset(0.0, 0.0));
      Offset offsetCenter = size.center(new Offset(0.0, -4.0));
      Offset renderOffset = new Offset(offsetCenter.dx, offsetCenter.dy*controller.value);
      canvas.drawLine(
      offsetTopCenter, 
      renderOffset, 
      lineStroke);
    }
    else {
      Offset offsetTopCenter = size.topCenter(new Offset(0.0, 0.0));
      Offset offsetBottom = size.bottomCenter(new Offset(0.0, 0.0));
      Offset renderOffset = new Offset(offsetBottom.dx, offsetBottom.dy*controller.value);
      canvas.drawLine(
      offsetTopCenter, 
      renderOffset, 
      lineStroke);
    }
    
    
    Paint circleFill = new Paint()
    ..color = lineColor
    ..style = PaintingStyle.fill;

    canvas.drawCircle(size.center(new Offset(0.0, -8.0)), 6.0, circleFill);

  }

  @override
  bool shouldRepaint(TimelinePainter oldDelegate) {
    return oldDelegate.lineColor!=lineColor || oldDelegate.backgroundColor!=backgroundColor;
  }

}

lib/timeline_element.dart
/* Copyright 2018 Rejish Radhakrishnan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

import 'package:flutter/material.dart';
import 'package:timeline/model/timeline_model.dart';
import 'timeline_painter.dart';

class TimelineElement extends StatelessWidget {

  final Color lineColor;
  final Color backgroundColor;
  final TimelineModel model;
  final bool firstElement;
  final bool lastElement;
  final Animation<double> controller;
  final Color headingColor;
  final Color descriptionColor;

  TimelineElement({
    @required this.lineColor,
    @required this.backgroundColor,
    @required this.model,
    this.firstElement = false,
    this.lastElement = false,
    this.controller, 
    this.headingColor, 
    this.descriptionColor
  });

  Widget _buildLine(BuildContext context, Widget child) {
    return new Container(
      width: 40.0,
      child: new CustomPaint(
        painter: new TimelinePainter(
          lineColor: lineColor,
          backgroundColor: backgroundColor,
          firstElement: firstElement,
          lastElement: lastElement,
          controller: controller
        ),
      ),
    );
  }

  Widget _buildContentColumn(BuildContext context) {
    return new Column(
      mainAxisSize: MainAxisSize.max,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        new Container(
          padding: const EdgeInsets.only(bottom: 8.0, top: 8.0),
          child: new Text(
            model.title.length>47?model.title.substring(0,47)+"...":model.title,
            style: new TextStyle(
              fontWeight: FontWeight.bold,
              color: headingColor!=null?headingColor:Colors.black,
            ),
          ),
        ),
        new Expanded(
                  child: new Text(
            model.description!=null?(model.description.length>50?model.description.substring(0,50)+"...":model.description):"", // To prevent overflowing of text to the next element, the text is truncated if greater than 75 characters
            style: new TextStyle(
              color: descriptionColor!=null?descriptionColor:Colors.grey,
            ),
          ),
        )
      ],
    );
  }

  Widget _buildRow(BuildContext context)
  {
    return new Container(
      height: 80.0,
      color: backgroundColor,
      padding: const EdgeInsets.symmetric(horizontal: 12.0, vertical: 0.0),
      child: new Row(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          new AnimatedBuilder(
            builder: _buildLine,
            animation: controller,
          ),
          new Expanded(
            child: _buildContentColumn(context),
            ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return _buildRow(context);
  }

}

lib/timeline_painter.dart
/* Copyright 2018 Rejish Radhakrishnan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

import 'dart:ui';
import 'package:flutter/material.dart';

class TimelinePainter extends CustomPainter {

  final Color lineColor;
  final Color backgroundColor;
  final bool firstElement;
  final bool lastElement;
  final Animation<double> controller;
  final Animation<double> height;

  TimelinePainter({
    @required this.lineColor,
    @required this.backgroundColor,
    this.firstElement = false,
    this.lastElement = false,
    this.controller
  }) :height = new Tween(begin: 0.0, end: 1.0).animate(new CurvedAnimation(
          parent: controller,
          curve: new Interval(
            0.45, 1.0,
            curve: Curves.ease),
        ),
      ),
      super(repaint: controller);

  @override
  void paint(Canvas canvas, Size size) {
    _centerElementPaint(canvas, size);
  }

  void _centerElementPaint(Canvas canvas, Size size)
  {
    Paint lineStroke = new Paint()
    ..color = lineColor
    ..strokeCap = StrokeCap.round
    ..strokeWidth = 2.0
    ..style = PaintingStyle.stroke;
    if(firstElement && lastElement)
    {
      // Do nothing
    }
    else if(firstElement)
    {
      Offset offsetCenter = size.center(new Offset(0.0, -4.0));
      Offset offsetBottom = size.bottomCenter(new Offset(0.0, 0.0));
      Offset renderOffset = new Offset(offsetBottom.dx, offsetBottom.dy*(0.5+(controller.value/2)));
      canvas.drawLine(
      offsetCenter, 
      renderOffset, 
      lineStroke);
    }
    else if(lastElement)
    {
      Offset offsetTopCenter = size.topCenter(new Offset(0.0, 0.0));
      Offset offsetCenter = size.center(new Offset(0.0, -4.0));
      Offset renderOffset = new Offset(offsetCenter.dx, offsetCenter.dy*controller.value);
      canvas.drawLine(
      offsetTopCenter, 
      renderOffset, 
      lineStroke);
    }
    else {
      Offset offsetTopCenter = size.topCenter(new Offset(0.0, 0.0));
      Offset offsetBottom = size.bottomCenter(new Offset(0.0, 0.0));
      Offset renderOffset = new Offset(offsetBottom.dx, offsetBottom.dy*controller.value);
      canvas.drawLine(
      offsetTopCenter, 
      renderOffset, 
      lineStroke);
    }
    
    
    Paint circleFill = new Paint()
    ..color = lineColor
    ..style = PaintingStyle.fill;

    canvas.drawCircle(size.center(new Offset(0.0, -8.0)), 6.0, circleFill);

  }

  @override
  bool shouldRepaint(TimelinePainter oldDelegate) {
    return oldDelegate.lineColor!=lineColor || oldDelegate.backgroundColor!=backgroundColor;
  }

}

lib/timeline-element.dart
/* Copyright 2018 Rejish Radhakrishnan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

import 'package:flutter/material.dart';
import 'package:meta/meta.dart';
import 'package:timeline/model/timeline-model.dart';
import 'timeline-painter.dart';

class TimelineElement extends StatelessWidget {

  final Color lineColor;
  final Color backgroundColor;
  final TimelineModel model;
  final bool firstElement;
  final bool lastElement;
  final Animation<double> controller;
  final Color headingColor;
  final Color descriptionColor;

  TimelineElement({
    @required this.lineColor,
    @required this.backgroundColor,
    @required this.model,
    this.firstElement = false,
    this.lastElement = false,
    this.controller, 
    this.headingColor, 
    this.descriptionColor
  });

  Widget _buildLine(BuildContext context, Widget child) {
    return new Container(
      width: 40.0,
      child: new CustomPaint(
        painter: new TimelinePainter(
          lineColor: lineColor,
          backgroundColor: backgroundColor,
          firstElement: firstElement,
          lastElement: lastElement,
          controller: controller
        ),
      ),
    );
  }

  Widget _buildContentColumn(BuildContext context) {
    return new Column(
      mainAxisSize: MainAxisSize.max,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        new Container(
          padding: const EdgeInsets.only(bottom: 8.0, top: 8.0),
          child: new Text(
            model.title.length>47?model.title.substring(0,47)+"...":model.title,
            style: new TextStyle(
              fontWeight: FontWeight.bold,
              color: headingColor!=null?headingColor:Colors.black,
            ),
          ),
        ),
        new Expanded(
                  child: new Text(
            model.description!=null?(model.description.length>50?model.description.substring(0,50)+"...":model.description):"", // To prevent overflowing of text to the next element, the text is truncated if greater than 75 characters
            style: new TextStyle(
              color: descriptionColor!=null?descriptionColor:Colors.grey,
            ),
          ),
        )
      ],
    );
  }

  Widget _buildRow(BuildContext context)
  {
    return new Container(
      height: 80.0,
      color: backgroundColor,
      padding: const EdgeInsets.symmetric(horizontal: 12.0, vertical: 0.0),
      child: new Row(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          new AnimatedBuilder(
            builder: _buildLine,
            animation: controller,
          ),
          new Expanded(
            child: _buildContentColumn(context),
            ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return _buildRow(context);
  }

}